File: Implementation/python/16926.py
import sys
input = sys.stdin.readline
n,m,r = map(int,input().split())
arr = [list(map(int,input().rstrip().split())) for _ in range(n)]
moves = [[1,0],[0,1],[-1,0],[0,-1]]

def rotate (i):
  prev = arr[i][i]
  for j in range(i+1,n-i):
    arr[j][i], prev = prev, arr[j][i]
  for j in range(i+1,m-i):
    arr[n-1-i][j], prev = prev, arr[n-1-i][j]
  for j in range(i+1,n-i):
    arr[n-1-j][m-1-i],prev = prev, arr[n-1-j][m-1-i]
  for j in range(i+1,m-i):
    arr[i][m-1-j],prev = prev, arr[i][m-1-j]

def solution(r):
  for i in range(r):
    for j in range(min(n,m)//2):
      rotate(j)

solution(r)
print("\n".join(" ".join(list(map(str,arr[i]))) for i in range(n)))


# 매우 느린 풀이
# import sys
# input = sys.stdin.readline
# n,m,r = map(int,input().split())
# arr = [list(map(int,input().rstrip().split())) for _ in range(n)]
# moves = [[1,0],[0,1],[-1,0],[0,-1]]

# def rotate (i,visited):
#   r=c=i
#   d = 0
#   prev = arr[r][c]
#   while d<4:
#     move = moves[d]
#     nr,nc = move[0]+r, move[1]+c
#     if nr<0 or nc<0 or nr>=n or nc>=m:
#       d += 1
#       continue
#     if visited[nr][nc]:
#       d += 1
#       continue
#     visited[nr][nc] = True
#     arr[nr][nc],prev = prev,arr[nr][nc]
#     r,c = nr,nc
# def solution(cnt):
#   if cnt ==0:
#     return
#   visited =[[False for _ in range(m)] for _ in range(n)]
#   for i in range(min(n,m)//2):
#     rotate(i,visited)
#   solution(cnt-1)
# solution(r)

# print("\n".join(" ".join(list(map(str,arr[i]))) for i in range(n)))

File: Implementation/javascript/16926.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
const [n, m, r] = input[0].split(" ").map(Number);
const arr = [];
for (let i = 1; i < input.length - 1; i++) {
  arr.push(input[i].split(" ").map(Number));
}

const solution = (n, m, r) => {
  const cnt = ~~(Math.min(n, m) / 2);
  const rotate = (i) => {
    let prev = arr[i][i];
    for (let j = i + 1; j < n - i; j++) {
      [arr[j][i], prev] = [prev, arr[j][i]];
    }
    for (let j = i + 1; j < m - i; j++) {
      [arr[n - 1 - i][j], prev] = [prev, arr[n - 1 - i][j]];
    }
    for (let j = i + 1; j < n - i; j++) {
      [arr[n - 1 - j][m - 1 - i], prev] = [prev, arr[n - 1 - j][m - 1 - i]];
    }
    for (let j = i + 1; j < m - i; j++) {
      [arr[i][m - 1 - j], prev] = [prev, arr[i][m - 1 - j]];
    }
  };
  for (let i = 0; i < cnt; i++) {
    let rotateCnt = new Array(r % ((n - i * 2) * 2 + (m - i * 2) * 2 - 4)).fill(0);
    rotateCnt.forEach(() => {
      rotate(i);
    });
  }
};
solution(n, m, r);
const answer = arr.reduce((acc, cur) => {
  return acc + cur.reduce((row, num) => row + num + " ", "") + "\n";
}, "");
console.log(answer);
